### ISSUE TITLE
clipping of sprites on running vehicles

### ISSUE THREAD

**Issue by strongbowe1** [OpenLoco/OpenLoco#1230]
I model and create my own vehicles and this has been bugging me for decades. It happens not just with my vehicles, but with stock ones too. The vehicles - especially trains - have very strange clipping and sometimes the ordering of the sprites is messed up, when running through curves and on inclines/declines.

![XWPwwUp](https://user-images.githubusercontent.com/37986589/154850035-58ff86e8-f5cd-4ea3-a5a9-98bf8a0c098e.png)

If this could be somehow fixed, I would be happy to see my creations run without flickering.

**Comment by LeftofZen** [comment]
Do you have any examples of it happening with stock vehicles?

**Comment by strongbowe1** [comment]
![clipping](https://user-images.githubusercontent.com/37986589/154013535-2d54e19f-fe57-4243-8744-03b74343caed.png)

It happens mainly with longer vehicles - this is the game default Century locomotive. Most mods for Locomotion are of real size vehicles, so the clipping is considerably worse than on the game default shortened/cartoonized vehicles.

**Comment by strongbowe1** [comment]
Here as well - Swiss Alps scenario, game standard passenger carriages on a steep narrow gauge track.

![clipping_swissalps](https://user-images.githubusercontent.com/37986589/154849334-40315a54-21c7-4adc-95d3-f2fe4018f134.png)

**Comment by strongbowe1** [comment]
And this is how my own, slightly longer, Glacier Express carriages look like on a steep narrow gauge track. In addition to the carriages clipping through each other, the front bogies are also visible through the carriage
![glacier](https://user-images.githubusercontent.com/37986589/154849966-8624d500-beb0-4b35-84b1-9b185fa65008.png)
.

**Comment by AaronVanGeffen** [comment]
Just so we understand correctly, does this clipping occur in vanilla Locomotion as well? That is, it's not a bug in OpenLoco, is it?

We have resolved some clipping artefacts in OpenRCT2 (same engine), so it's not unlikely that we can eventually do the same for OpenLoco. We're not at that point of our implementation yet, though.

**Comment by strongbowe1** [comment]
Yes, it's an age-old Locomotion bug, not OpenLoco. It would be great if it could be fixed with OpenLoco though, especially for mods of longer vehicles like mine.

**Comment by LeftofZen** [comment]
I think the issue is with the objects, not the rendering - the vehicles are too close together - including vanilla ones. I think adding a couple of units of spacing here will help resolve these issues.

Similar to the recent #2678, #2683

**Comment by strongbowe1** [comment]
Yes, adding spacing will help, but it'll look strange to have the carriages in a train spread so far apart. Also, the longer the object, the more severe the clipping issues are - just like with the vanilla ones.

And some issues can't even be fixed by spreading the vehicles like the bogies showing through the vehicle itself when it's too long.

But again, not a big issue for myself as I've lived with it since the game was released. But worth pointing out in my view.

**Comment by duncanspumpkin** [comment]
I've noticed that even some vanilla objects are longer than the max supported length in the engine so it's not too surprising this is happening. It would be interesting to see if this still happens if we enable the ship and airplane sprite limits for trains as well.

**Comment by glenjimen** [comment]
Looking deeper into this issue, I think the vehicle sprites are sometimes clipping, and sometimes being drawn in the wrong order.  On diagonals, long sprites clip. 

<img width="629" height="152" alt="Image" src="https://github.com/user-attachments/assets/cf1dfa87-b3c2-4d9e-9d76-a6fd298d0a48" />

It's almost like the adjacent track sprite is being drawn over the car body, and sometimes the car bogie.  You can see it in the ghost mouse-over preview.  It would make more sense if the ground tile behind the car was being drawn over-top the car, causing the clipping, but then we would see a diagonal edge to the sprite, not a vertical edge.  And you can see the clipping occurs at the edge of a diagonal track tile, which leads me to believe that is the culprit.  

https://github.com/user-attachments/assets/0d6d49d7-8abd-457c-ab3d-f781dc740608

A separate but similar issue is draw order on curves, specifically tight curves. 

<img width="416" height="206" alt="Image" src="https://github.com/user-attachments/assets/d7b710a5-0e46-4802-9db6-b0f06da2a3db" />

Here vehicle sprites on the 'blue' curve are being drawn over vehicle sprites on the 'red' curves.  There's no hard edge where the background shows thru, it's just sprite overlap.  And this is with very short vehicle sprites.  Longer ones overlap as well.

<img width="431" height="181" alt="Image" src="https://github.com/user-attachments/assets/a543f787-27d8-4ddf-8a39-2ed6048d2d69" />

These are all original bugs, not introduced by OpenLoco.  Is it possible to change the draw order of vehicle sprites on curves?
